### ISSUE TITLE
Failed to decode uncompressed block (length check)

### ISSUE THREAD

**Issue by agu-z** [folkertdev/elm-flate#4]
Hi!

#### Context

I'm writing a library to read `xlsx` spreadsheets from Elm, and as you might know, `xlsx` files are zipped XML documents. 

So, [I wrote a lib](https://package.elm-lang.org/packages/agu-z/elm-zip/latest/Zip) that can read the zip file format, extract the compressed data, and inflate it with `elm-flate`. I tried it with several archives made by different programs, and it worked great! 

#### Problem

Unfortunately, I haven't been able to inflate some files deflated by Excel. I can read the overall ZIP structure and get the compressed data, but `Flate.inflate` returns `Nothing` for most files in the archive.

To confirm that I didn't have a corrupted archive, I tried unzipping the spreadsheet with the `unzip` command, and it worked fine. I also made sure that the compression method was 8 (Deflate).

I cloned this repo and started playing in the REPL until I found that a block in my test data is failing this check:
https://github.com/folkertdev/elm-flate/blob/60fd8ede44838f055f92c2f678bddcc71c9ac01b/src/Inflate/Internal.elm#L540-L543

Apparently, at that point, `length` is `0x0000` while `invLength` is `0xFFFC` instead of `0xFFFF`. 

I thought that maybe I was reading the deflate data from the wrong offset. So, I wrote the bytes to a file and tried inflating it using zlib directly from Python:
```python
>>> import zlib
>>> zlib.decompress(open('test.bin', 'rb').read(), -8) # -8 indicates no header/trailer
'<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<sst xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main" count="6" uniqueCount="6"><si><t>JK</t></si><si><t>PK</t></si><si><t>Initials</t></si><si><t>Salary</t></si><si><t>LP</t></si><si><t>Start</t></si></sst>'
```

It inflated properly. Then I looked at `zlib`'s source code and found that it appears to perform the same check:
https://github.com/madler/zlib/blob/cacf7f1d4e3d44d871b605da3b647f07d718623f/inflate.c#L890

Is it possible that the library is incorrectly detecting an uncompressed block or trying to read it from the wrong offset?

#### Data

<details>
<summary>
REPL Example
</summary>

```elm
> import Inflate.Internal
> import Bytes.Encode as E
>
> data = [100, 143, 65, 107, 2, 49, 16, 133, 239, 66, 255, 67, 152, 123, 205, 218, 131, 20, 73, 226, 161, 80, 208, 246, 32, 84, 127, 192, 176, 59, 186, 129, 100, 178, 102, 102, 75, 253, 247, 221, 82, 68, 208, 227, 247, 189, 247, 14, 207, 173, 127, 114, 50, 223, 84, 37, 22, 246, 176, 152, 55, 96, 136, 219, 210, 69, 62, 121, 56, 236, 223, 159, 95, 193, 136, 34, 119, 152, 10, 147, 135, 11, 9, 172, 195, 211, 204, 137, 168, 153, 182, 44, 30, 122, 213, 97, 101, 173, 180, 61, 101, 148, 121, 25, 136, 167, 228, 88, 106, 70, 157, 176, 158, 172, 12, 149, 176, 147, 158, 72, 115, 178, 47, 77, 179, 180, 25, 35, 131, 105, 203, 200, 234, 97, 9, 102, 228, 120, 30, 233, 237, 202, 193, 73, 12, 78, 195, 246, 195, 89, 13, 206, 254, 209, 191, 217, 61, 152, 13, 71, 141, 152, 228, 190, 249, 133, 9, 235, 229, 222, 126, 238, 30, 122, 138, 85, 111, 210, 78, 215, 194, 47, 0, 0, 0, 255, 255, 3, 0]
>
> data |> List.map E.unsignedInt8 |> E.sequence |> E.encode |> Inflate.Internal.inflate
Err ("inflateUncompressedBlock: ran out of bounds")
```
</details>

Here's also a [test-sample.zip](https://github.com/folkertdev/elm-flate/files/5860535/test-sample.zip) that contains an example workbook and a python3 script that reads the raw data from the archive and prints out a list of raw bytes to use with `elm-flate`. 

#### --

I'm happy to dig deeper into this issue, but I would appreciate some direction since I'm no Deflate expert.

Thank you so much for making this library! Having a deflate implementation in pure Elm is glorious :heart:

**Comment by folkertdev** [comment]
thanks for the descriptive issue!

I've confirmed that there is in fact an issue somewhere in `inflate` where it looks like some bits are flipped/lost. Now I need to actually find where that problem is introduced...

**Comment by folkertdev** [comment]
just published version 2.0.5 which fixes this issue. Let me know if you run into any more trouble.

The problem was that I assumed that new blocks start on byte boundaries, but that is not true in Deflate. However, when reading uncompressed blocks we must skip to the next byte boundary. Three bits are read already, so I read another 5 bits. But clearly (in hindsight) this 5-bit jump would only put you on a byte boundary if you initially were on one (before reading those first 3 bits).

So the fix is simple: dynamically determine how many bits to read until you reach a byte boundary (multiple of 8 bytes remaining) and skip that many bits forward.

Again thanks for reporting this. This kind of package gets more stable one bugfix at a time.

**Comment by agu-z** [comment]
Oh I see, that's very interesting. I was trying to deduce what that 5-bit jump was about, now it makes total sense.

Thank you for providing a bugfix so quickly :)
